cmake/modules/boost_asio_static_variables.patch

```diff
@@ -0,0 +1,18 @@
+--- a/boost/asio/detail/call_stack.hpp	2020-11-09 15:36:10.800244119 -0500
++++ b/boost/asio/detail/call_stack.hpp	2020-11-09 15:37:36.199826023 -0500
+@@ -114,9 +114,12 @@
+   static tss_ptr<context> top_;
+ };
+ 
+-template <typename Key, typename Value>
+-tss_ptr<typename call_stack<Key, Value>::context>
+-call_stack<Key, Value>::top_;
++// moving static variables to ceph-common to avoid crash due to multiple
```

Comment:
@tchaikov I wanted to get your input on this first commit. Short story is that `boost::asio` doesn't play well as a header-only library when used across shared libraries since `libceph-common`, `librados`, and `librbd` all had their own instance of these static member variables. In this case, `librados` owns the reactor and `librbd` re-uses it to avoid the extra threads and the unnecessary context switching but that resulted in random crashes deep within the scheduler due to a single thread switching between shared library data segments in the same call stack.

Reply:
@dillaman your analysis makes sense to me. FWIW, i have a different observation in my environment.

```
$ nm -C lib/libceph-common.so | grep 'top_' | grep 'boost::asio::detail::call_stack'
0000000002606e30 u guard variable for boost::asio::detail::call_stack<boost::asio::detail::strand_service::strand_impl, unsigned char>::top_
0000000002606e28 u guard variable for boost::asio::detail::call_stack<boost::asio::detail::thread_context, boost::asio::detail::thread_info_base>::top_
0000000002606e40 u guard variable for boost::asio::detail::call_stack<boost::asio::detail::strand_executor_service::strand_impl, unsigned char>::top_
0000000002606e09 u boost::asio::detail::call_stack<boost::asio::detail::strand_service::strand_impl, unsigned char>::top_
0000000002606e08 u boost::asio::detail::call_stack<boost::asio::detail::thread_context, boost::asio::detail::thread_info_base>::top_
0000000002606e0b u boost::asio::detail::call_stack<boost::asio::detail::strand_executor_service::strand_impl, unsigned char>::top_

$ nm -C lib/librados.so | grep 'top_' | grep 'boost::asio::detail::call_stack'
00000000003a6850 b guard variable for boost::asio::detail::call_stack<boost::asio::detail::strand_service::strand_impl, unsigned char>::top_
00000000003a6848 b guard variable for boost::asio::detail::call_stack<boost::asio::detail::thread_context, boost::asio::detail::thread_info_base>::top_
00000000003a6860 b guard variable for boost::asio::detail::call_stack<boost::asio::detail::strand_executor_service::strand_impl, unsigned char>::top_
00000000003a6829 b boost::asio::detail::call_stack<boost::asio::detail::strand_service::strand_impl, unsigned char>::top_
00000000003a6828 b boost::asio::detail::call_stack<boost::asio::detail::thread_context, boost::asio::detail::thread_info_base>::top_
00000000003a682b b boost::asio::detail::call_stack<boost::asio::detail::strand_executor_service::strand_impl, unsigned char>::top_

$ nm -C lib/librbd.so | grep 'top_' | grep 'boost::asio::detail::call_stack'
0000000000b9b7d0 u guard variable for boost::asio::detail::call_stack<boost::asio::detail::strand_service::strand_impl, unsigned char>::top_
0000000000b9b7c8 u guard variable for boost::asio::detail::call_stack<boost::asio::detail::thread_context, boost::asio::detail::thread_info_base>::top_
0000000000b9b7e0 u guard variable for boost::asio::detail::call_stack<boost::asio::detail::strand_executor_service::strand_impl, unsigned char>::top_
0000000000b9b7a9 u boost::asio::detail::call_stack<boost::asio::detail::strand_service::strand_impl, unsigned char>::top_
0000000000b9b7a8 u boost::asio::detail::call_stack<boost::asio::detail::thread_context, boost::asio::detail::thread_info_base>::top_
0000000000b9b7ab u boost::asio::detail::call_stack<boost::asio::detail::strand_executor_service::strand_impl, unsigned char>::top_
```

so these `top_` variables of the instantiated templates are 

1. local 
1. defined in librados

i am using GCC 10.2.0 and GNU Binutils 2.35.1 .

also, if we choose to use the reactor owned by librados (or libneorados?), why not let it contain the `top_` symbols instead?

Reply:
That was actually my first attempt at fixing it. `libceph-common.so` and `librbd.so` had global-unique versions of those symbols "`u`" but `librados.so` had a local bss. I fixed `librados` by adding those symbols to the `librados.map` to switch those to global unique and that seemed to work in combination w/ disabling the `__thread` TLS prefix for `boost::asio`, but then I read it was a GNU extension to ELF so I wasn't sure how that would work on other platforms / compilers.

Reply:
(of course it looks like the builders are caching boost so that makes things more difficult since I need to patch it).

Reply:
Perhaps due to the "system" boost issue I should just revert to my original:

```
diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index b8110eab621..bfb5c0e8674 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -28,7 +28,8 @@ add_definitions(
   -D_REENTRANT
   -D_THREAD_SAFE
   -D__STDC_FORMAT_MACROS
-  -D_FILE_OFFSET_BITS=64)
+  -D_FILE_OFFSET_BITS=64
+  -DBOOST_ASIO_DISABLE_THREAD_KEYWORD_EXTENSION)
 if(LINUX)
   add_definitions("-D_GNU_SOURCE")
 endif()
diff --git a/src/librados/librados.map b/src/librados/librados.map
index b7552be6111..279a0ba0691 100644
--- a/src/librados/librados.map
+++ b/src/librados/librados.map
@@ -1,4 +1,14 @@
 LIBRADOS_PRIVATE {
+       global:
+               extern "C++" {
+                       "guard variable for boost::asio::detail::call_stack<boost::asio::detail::strand_executor_service::strand_impl, unsigned char>::top_";
+                       "guard variable for boost::asio::detail::call_stack<boost::asio::detail::strand_service::strand_impl, unsigned char>::top_";
+                       "guard variable for boost::asio::detail::call_stack<boost::asio::detail::thread_context, boost::asio::detail::thread_info_base>::top_";
+                       "boost::asio::detail::call_stack<boost::asio::detail::strand_executor_service::strand_impl, unsigned char>::top_";
+                       "boost::asio::detail::call_stack<boost::asio::detail::strand_service::strand_impl, unsigned char>::top_";
+                       "boost::asio::detail::call_stack<boost::asio::detail::thread_context, boost::asio::detail::thread_info_base>::top_";
+
+               };
        local:  *;
 };
```

Reply:
yeah, i packaged the system boost packages. but to avoid further issues when user tries to build ceph with the boost shipped by distro. i agree that the change above is a better solution.